### PR TITLE
feat(streaming): apply StateTable to hash join

### PR DESCRIPTION
**Pull request by yuhao-su**
## What's changed and what's your intention?

- apply StateTable to hash join
- solve unflushed degree issue

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#2794 
#2795

### CODE REVIEW

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 1]
```diff
@@ -275,49 +233,52 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
+    }
+
+    /// Insert a key
+    pub fn insert(&mut self, join_key: &K, pk: Row, value: JoinRow) -> RwResult<()>{
+        if let Some(entry) = self.inner.get_mut(join_key) {
+            entry.insert(pk, value.clone());
+        }
+        let key = join_key.clone().deserialize(self.join_key_data_types.iter())?;
+        // If no cache maintained, only update the flush buffer.
+        self.state_table.insert(&key, value.into_row())?;
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/join/mod.rs:264]: So the row count currently is not serialized, right?

Well the e2e tests shouldn't be able to pass without resolving this...? I guess we should decide on the way soon.

Comment (yuhao-su) [reply src/stream/src/executor/managed_state/join/mod.rs]: The row count is currently added to the tail of a row as a datum as a workaround. :>

File: src/stream/src/executor/hash_join.rs [hunk 2]
```diff
@@ -139,8 +142,9 @@ impl<K: HashKey, S: StateStore> std::fmt::Debug for JoinSide<K, S> {
 }
 
 impl<K: HashKey, S: StateStore> JoinSide<K, S> {
+    // WARNING: Please do not call this until we implement it.
     fn is_dirty(&self) -> bool {
-        self.ht.values().any(|state| state.is_dirty())
+        false
```
Comment (BowenXiao1999) [inline src/stream/src/executor/hash_join.rs:147]: Should panic here?

**Review by BowenXiao1999 — COMMENTED** [review]
It seems this PR consists of mutiple changes:
* `Update` on mem_table.
* Integrate StateTable
etc

File: src/stream/src/executor/hash_join.rs (hunk 2 above)
Comment (yuhao-su) [reply src/stream/src/executor/hash_join.rs:147]: Maybe implement it after #2999. So we can get if mem table is empty

File: src/storage/src/table/mem_table.rs [hunk 3]
```diff
@@ -108,6 +108,36 @@ impl MemTable {
...
+                RowOp::Delete(_) => {
+                    panic!(
+                        "invalid flush status: double delete {:?} -> {:?}",
+                        e.key(),
+                        old_value
+                    );
+                }
+                x @ RowOp::Update(_) => {
+                    if let RowOp::Update(ref mut value) = x {
+                        let (original_old_value, original_new_value) = std::mem::take(value);
```
Comment (BugenZhao) [inline src/storage/src/table/mem_table.rs:131]: Why not directly match `RowOp::Update(ref mut value)` here? Same for `delete`.

**Comment by yuhao-su** [comment]
> It seems this PR consists of mutiple changes:
> 
> * `Update` on mem_table.
> * Integrate StateTable
>   etc

Need to fix #2794 or test fails. To do that we must have update.

File: src/common/src/array/data_chunk_iter.rs [hunk 4]
```diff
@@ -315,6 +315,13 @@ impl Row {
         }
         Ok(HashCode(hasher.finish()))
     }
+
+    /// Get an owned `Row` by the given `indices` from current row.
+    ///
+    /// Use `datum_refs_by_indices` if possible instead to avoid allocating owned datums.
+    pub fn by_indices(&self, indices: &[usize]) -> Row {
```
Comment (BugenZhao) [inline src/common/src/array/data_chunk_iter.rs:322]: Off the topic: We should really review the current implementation of `serialize` functions. In most cases there's no need to allocate a `Row`.

File: src/stream/src/executor/managed_state/join/join_entry_state.rs [hunk 5]
```diff
@@ -32,205 +22,54 @@ type JoinEntryStateValues<'a> = btree_map::Values<'a, PkType, StateValueType>;
...
 
-/// We manages a `BTreeMap` in memory for all entries belonging to a join key,
-/// since each `WriteBatch` is an ordered list of key-value pairs.
-/// When evicted, `BTreeMap` does not hold any entries.
-pub struct JoinEntryState<S: StateStore> {
+/// We manages a `BTreeMap` in memory for all entries belonging to a join key.
+/// When evicted, `cached` does not hold any entries.
+/// If a `JoinEntryState` exists for a join key, the all records under this
+/// join key will be presented in the cache.
+pub struct JoinEntryState {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/join_entry_state.rs:29]: Seems this struct is definitely a `BTreeMap` now. 🤣 How about implement `Deref` for it?

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 6]
```diff
@@ -14,25 +14,30 @@
...
+use risingwave_common::error::{ErrorCode, Result as RwResult, RwError};
 use risingwave_common::hash::{HashKey, PrecomputedBuildHasher};
-use risingwave_common::types::{DataType, Datum};
+use risingwave_common::types::{DataType, Datum, ScalarImpl};
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_storage::table::state_table::StateTable;
 use risingwave_storage::{Keyspace, StateStore};
 use stats_alloc::{SharedStatsAlloc, StatsAlloc};
 
+type DegreeType = i64;
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:35]: Suggest still using unsigned integer and cast it when turn it into a datum.

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 7]
```diff
@@ -146,29 +136,47 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
             .iter()
-            .map(|idx| data_types[*idx].clone())
+            .enumerate()
+            .map(|(id, data_type)| ColumnDesc::unnamed(ColumnId::new(id as i32), data_type.clone()))
             .collect_vec();
+
+        let table_pk_indices = [join_key_indices, pk_indices.clone()].concat();
+
+        // Order type doesn't matter here. Arbitrarily choose one.
+        let order_types = vec![OrderType::Descending; table_pk_indices.len()];
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:160]: I'm not sure whether `Ascending` can be faster to serialize due to no need of toggling bits.

File: src/stream/src/executor/managed_state/join/mod.rs (hunk 1 above)
Comment (BowenXiao1999) [reply src/stream/src/executor/managed_state/join/mod.rs]: What's your terminology here? row count eqauls to match degree in #2794 ?

**Review by BowenXiao1999 — COMMENTED** [review]
Overall LGTM. The hash agg is more natural to integrate with StateTable so the logic looks clean.

But indeed this PR can be divided (?).

File: proto/stream_plan.proto [hunk 8]
```diff
@@ -85,18 +85,19 @@ message HashJoinNode {
...
-  bool is_delta_join = 6;
+  bool is_delta_join = 5;
   // Used for internal table states. Id of the left table.
-  uint32 left_table_id = 7;
+  uint32 left_table_id = 6;
   // Used for internal table states. Id of the right table.
-  uint32 right_table_id = 8;
+  uint32 right_table_id = 7;
+  repeated uint32 dist_key_l = 8;
+  repeated uint32 dist_key_r = 9;
```
Comment (BowenXiao1999) [inline proto/stream_plan.proto:97]: dist key of hash join is not join key?

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 9]
```diff
@@ -184,140 +192,129 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
-        );
+    /// Insert a [`JoinEntryState`]
+    pub fn insert_state(&mut self, key: &K, state: JoinEntryState) {
         self.inner.put(key.clone(), state);
+    }
+
+    pub fn inc_degree(&mut self, join_row: &mut JoinRow) -> RwResult<()> {
+        let old_row = join_row.clone().into_row();
+        join_row.inc_degree();
+        let new_row = join_row.clone().into_row();
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:300]: This could be too costly. 😢 Let's optimize it later.

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM.

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 10]
```diff
@@ -184,140 +192,129 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
-        Ok(self.keyspace.append(key_encoded))
-    }
-
     /// Returns a mutable reference to the value of the key in the memory, if does not exist, look
-    /// up in remote storage and return, if still not exist, return None.
+    /// up in remote storage and return. If not exist in remote storage, a
+    /// `JoinEntryState` with empty cache will be returned.
     // #[allow(dead_code)]
-    // pub async fn get<'a, 'b>(&'a mut self, key: &'b K) -> Option<&'a HashValueType<S>> {
+    // pub async fn get<'a>(&'a mut self, key: &K) -> Option<&'a HashValueType> {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:199]: Let's mark it with `#[cfg(any())]` or just remove it to keep it clean.

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 11]
```diff
@@ -184,140 +192,129 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
 
     /// Returns a mutable reference to the value of the key in the memory, if does not exist, look
-    /// up in remote storage and return, if still not exist, return None.
-    pub async fn get_mut(&mut self, key: &K) -> Option<&mut HashValueType<S>> {
-        let state = self.inner.get(key);
-        // TODO: we should probably implement a entry function for `LruCache`
+    /// up in remote storage and return. If not exist in remote storage, a
+    /// `JoinEntryState` with empty cache will be returned.
+    // #[allow(dead_code)]
+    // pub async fn get_mut<'a>(&'a mut self, key: &'a K) -> Option<&'a mut HashValueType> {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:216]: Ditto.

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**